src/snowflake/snowpark/session.py

```diff
@@ -917,6 +918,80 @@ def add_requirements(self, file_path: str) -> None:
+        that are not available in Snowflake, refer to :meth:`~snowflake.snowpark.Session.custom_package_usage_config`.
+
+        This function is **experimental**, please do not use it in production!
+
+        Example::
+
+            >>> from snowflake.snowpark.functions import udf
+            >>> import numpy
+            >>> import pandas
+            >>> # test_requirements.txt contains "numpy" and "pandas"
+            >>> session.custom_package_usage_config = {"enabled": True, "force_push": True} # Recommended configuration
+            >>> session.replicate_local_environment(ignore_packages={"snowflake-snowpark-python", "snowflake-connector-python", "urllib3", "tzdata", "numpy"})
+            >>> @udf
+            ... def get_package_name_udf() -> list:
+            ...     return [numpy.__name__, pandas.__name__]
+            >>> session.sql(f"select {get_package_name_udf.name}()").to_df("col1").show()
+            --------------
+            |"COL1"      |
+            --------------
+            |[           |
+            |  "numpy",  |
+            |  "pandas"  |
+            |]           |
+            --------------
+            <BLANKLINE>
+            >>> session.clear_packages()
+            >>> session.clear_imports()
+
+        Args:
+            reinstall_packages: Ignores environment present on persist_path and overwrites it with a fresh installation.
+            ignore_packages: Set of packages that will be ignored.
+            persist_path: A remote stage directory path where packages not present in Snowflake will be persisted. Mentioning
+             this path will speed up automated package loading.
+
+
+        Note:
+            1. This method will add packages for all UDFs created later in the current
+            session. If you only want to add packages for a specific UDF, you can use
+            ``packages`` argument in :func:`functions.udf` or
+            :meth:`session.udf.register() <snowflake.snowpark.udf.UDFRegistration.register>`.
+
+            2. We recommend you to `setup the local environment with Anaconda <https://docs.snowflake.com/en/developer-guide/udf/python/udf-python-packages.html#local-development-and-testing>`_,
+            to ensure the consistent experience of a UDF between your local environment
+            and the Snowflake server.
+        """
+        default_packages = ["wheel", "pip", "setuptools"]
+        ignore_packages = {} if ignore_packages is None else ignore_packages
+
+        packages = []
+        for package in pkg_resources.working_set:
+            if package.key in ignore_packages:
+                _logger.info(f"{package.key} found in environment, ignoring...")
+                continue
+            if package.key in default_packages:
+                _logger.info(f"{package.key} is available by default, ignoring...")
+                continue
+            packages.append(
+                f"{package.key}{'==' + package.version if package.has_version() else ''}"
```

Comment:
What if user's local version is not available on Snowflake? I think this will directly cause a failure 🤔

Reply:
based on auto-upload settings, it might get installed via pip, zipped and uploaded to a stage. Is that right Vivek?

Reply:
Yes, if the user's package versions are not on Snowflake and package auto-upload is not enabled, `_resolve_packages` will throw an error.

Reply:
Yeah that makes sense. I was just wondering if we should provide a relaxed mode, such that if:

- the user's environment contains packages that exists on Snowflake but not the exact version
- This package cannot be installed and uploaded to a stage

Then we just use the package without specifying the version. 

I'm not sure if this is a legit use case so just thinking out loud.

Reply:
Yeah maybe we can have a `relax=True` argument, sounds useful.